File: cms/src/seed/lobby-layout.ts
import type { Payload } from 'payload'

export async function seedLobbyLayout(payload: Payload): Promise<void> {
  console.log('Seeding lobby layout...')

  try {
    // Check if default layout already exists
    const existing = await payload.find({
      collection: 'lobby-layouts',
      where: {
        slug: { equals: 'web-default' },
      },
    })

    if (existing.docs.length > 0) {
      console.log('Default lobby layout already exists, skipping...')
      return
    }

    // Get hero promotions for carousel
    const heroPromotions = await payload.find({
      collection: 'promotions',
      where: {
        placement: { equals: 'hero' },
        status: { equals: 'live' },
      },
      sort: '-priority',
    })

    // Get banner promotion
    const bannerPromotions = await payload.find({
      collection: 'promotions',
      where: {
        placement: { equals: 'banner' },
        status: { equals: 'live' },
      },
      sort: '-priority',
      limit: 1,
    })

    // Build sections
    const sections: Record<string, unknown>[] = []

    // Hero Carousel Section
    if (heroPromotions.docs.length > 0) {
      sections.push({
        blockType: 'carousel-section',
        promotions: heroPromotions.docs.map((p) => p.id),
        autoPlay: true,
        autoPlayInterval: 5000,
        showDots: true,
        showArrows: true,
        height: 'large',
      })
    }

    // Suggested Games Section (Personalized)
    sections.push({
      blockType: 'suggested-games-section',
      title: 'Recommended for You',
      subtitle: 'Games picked just for you',
      mode: 'personalized',
      placement: 'homepage-suggested',
      limit: 10,
      fallbackToPopular: true,
      showScrollButtons: true,
      cardSize: 'medium',
    })

    // Popular Slots Grid
    sections.push({
      blockType: 'game-grid-section',
      title: 'Popular Slots',
      subtitle: 'Top rated by our players',
      filterType: 'type',
      gameType: 'slot',
      limit: 12,
      columns: '6',
      showMore: true,
      cardSize: 'medium',
      showJackpot: true,
      showProvider: true,
    })

    // Banner Section
    if (bannerPromotions.docs.length > 0) {
      sections.push({
        blockType: 'banner-section',
        promotion: bannerPromotions.docs[0].id,
        size: 'large',
        alignment: 'center',
        showCountdown: true,
        rounded: true,
        marginTop: 32,
        marginBottom: 32,
      })
    }

    // Live Casino Grid
    sections.push({
      blockType: 'game-grid-section',
      title: 'Live Casino',
      subtitle: 'Real dealers, real action',
      filterType: 'type',
      gameType: 'live',
      limit: 6,
      columns: '6',
      showMore: true,
      cardSize: 'medium',
      showJackpot: false,
      showProvider: true,
    })

    // Jackpot Games
    sections.push({
      blockType: 'game-grid-section',
      title: 'Jackpot Games',
      subtitle: 'Win big today',
      filterType: 'jackpot',
      limit: 6,
      columns: '6',
      showMore: true,
      cardSize: 'medium',
      showJackpot: true,
      showProvider: true,
    })

    // New Games Section
    sections.push({
      blockType: 'suggested-games-section',
      title: 'New Arrivals',
      subtitle: 'Fresh games added this week',
      mode: 'manual',
      limit: 8,
      showScrollButtons: true,
      cardSize: 'medium',
    })

    // Table Games Grid
    sections.push({
      blockType: 'game-grid-section',
      title: 'Table Games',
      subtitle: 'Classic casino favorites',
      filterType: 'type',
      gameType: 'table',
      limit: 6,
      columns: '6',
      showMore: true,
      cardSize: 'medium',
      showJackpot: false,
      showProvider: true,
    })

    // Instant Win Games
    sections.push({
      blockType: 'game-grid-section',
      title: 'Instant Win',
      subtitle: 'Quick games, instant fun',
      filterType: 'type',
      gameType: 'instant',
      limit: 6,
      columns: '6',
      showMore: true,
      cardSize: 'medium',
      showJackpot: false,
      showProvider: true,
    })

    // Create the layout
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    await payload.create({
      collection: 'lobby-layouts',
      data: {
        slug: 'web-default',
        name: 'Web Default Layout',
        platform: 'web',
        isDefault: true,
        sections,
      } as any,
    })

    console.log('Created default lobby layout')
  } catch (error) {
    console.error('Failed to create lobby layout:', error)
  }

  console.log('Lobby layout seeding completed!')
}

File: social_casino_app/lib/models/lobby_layout.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'game.dart';
import 'promotion.dart';

part 'lobby_layout.freezed.dart';
part 'lobby_layout.g.dart';

enum Platform {
  @JsonValue('web')
  web,
  @JsonValue('mobile')
  mobile,
}

enum CardSize {
  @JsonValue('small')
  small,
  @JsonValue('medium')
  medium,
  @JsonValue('large')
  large,
}

enum CarouselHeight {
  @JsonValue('small')
  small,
  @JsonValue('medium')
  medium,
  @JsonValue('large')
  large,
  @JsonValue('full')
  full,
}

enum BannerSize {
  @JsonValue('small')
  small,
  @JsonValue('medium')
  medium,
  @JsonValue('large')
  large,
}

enum FilterType {
  @JsonValue('manual')
  manual,
  @JsonValue('type')
  type,
  @JsonValue('tag')
  tag,
  @JsonValue('popular')
  popular,
  @JsonValue('new')
  newGames,
  @JsonValue('jackpot')
  jackpot,
  @JsonValue('featured')
  featured,
}

enum SuggestedMode {
  @JsonValue('manual')
  manual,
  @JsonValue('personalized')
  personalized,
}

enum DisplayStyle {
  @JsonValue('horizontal')
  horizontal,
  @JsonValue('grid')
  grid,
  @JsonValue('single-row')
  singleRow,
  @JsonValue('featured-left')
  featuredLeft,
  @JsonValue('featured-right')
  featuredRight,
  @JsonValue('featured-top')
  featuredTop,
}

enum BannerAlignment {
  @JsonValue('left')
  left,
  @JsonValue('center')
  center,
  @JsonValue('right')
  right,
}

@freezed
class CarouselSectionBlock with _$CarouselSectionBlock {
  const factory CarouselSectionBlock({
    @Default('carousel-section') String blockType,
    String? title,
    List<Promotion>? promotions,
    @Default(true) bool autoPlay,
    @Default(5000) int autoPlayInterval,
    @Default(true) bool showDots,
    @Default(true) bool showArrows,
    @Default(CarouselHeight.medium) CarouselHeight height,
  }) = _CarouselSectionBlock;

  factory CarouselSectionBlock.fromJson(Map<String, dynamic> json) =>
      _$CarouselSectionBlockFromJson(json);
}

@freezed
class SuggestedGamesSectionBlock with _$SuggestedGamesSectionBlock {
  const factory SuggestedGamesSectionBlock({
    @Default('suggested-games-section') String blockType,
    required String title,
    String? subtitle,
    required SuggestedMode mode,
    List<Game>? manualGames,
    required String placement,
    @Default(12) int limit,
    @Default(true) bool fallbackToPopular,
    @Default(true) bool showScrollButtons,
    @Default(CardSize.medium) CardSize cardSize,
  }) = _SuggestedGamesSectionBlock;

  factory SuggestedGamesSectionBlock.fromJson(Map<String, dynamic> json) =>
      _$SuggestedGamesSectionBlockFromJson(json);
}

@freezed
class GameGridSectionBlock with _$GameGridSectionBlock {
  const factory GameGridSectionBlock({
    @Default('game-grid-section') String blockType,
    required String title,
    String? subtitle,
    required FilterType filterType,
    List<Game>? manualGames,
    String? gameType,
    String? tag,
    @Default(DisplayStyle.horizontal) DisplayStyle displayStyle,
    @Default(2) int rows,
    Game? featuredGame,
    @Default(12) int limit,
    @Default('4') String columns,
    @Default(true) bool showMore,
    String? moreLink,
    @Default(CardSize.medium) CardSize cardSize,
    @Default(true) bool showJackpot,
    @Default(true) bool showProvider,
  }) = _GameGridSectionBlock;

  factory GameGridSectionBlock.fromJson(Map<String, dynamic> json) =>
      _$GameGridSectionBlockFromJson(json);
}

@freezed
class BannerSectionBlock with _$BannerSectionBlock {
  const factory BannerSectionBlock({
    @Default('banner-section') String blockType,
    Promotion? promotion,
    @Default(BannerSize.medium) BannerSize size,
    @Default(BannerAlignment.center) BannerAlignment alignment,
    @Default(false) bool showCountdown,
    @Default(true) bool rounded,
    @Default(16) int marginTop,
    @Default(16) int marginBottom,
  }) = _BannerSectionBlock;

  factory BannerSectionBlock.fromJson(Map<String, dynamic> json) =>
      _$BannerSectionBlockFromJson(json);
}

/// Union type for lobby sections - uses custom fromJson/toJson
sealed class LobbySectionBlock {
  const LobbySectionBlock();

  factory LobbySectionBlock.fromJson(Map<String, dynamic> json) {
    final blockType = json['blockType'] as String?;
    switch (blockType) {
      case 'carousel-section':
        return CarouselSection(CarouselSectionBlock.fromJson(json));
      case 'suggested-games-section':
        return SuggestedGamesSection(SuggestedGamesSectionBlock.fromJson(json));
      case 'game-grid-section':
        return GameGridSection(GameGridSectionBlock.fromJson(json));
      case 'banner-section':
        return BannerSection(BannerSectionBlock.fromJson(json));
      default:
        throw ArgumentError('Unknown block type: $blockType');
    }
  }

  Map<String, dynamic> toJson();
}

class CarouselSection extends LobbySectionBlock {
  final CarouselSectionBlock data;
  const CarouselSection(this.data);

  @override
  Map<String, dynamic> toJson() => data.toJson();
}

class SuggestedGamesSection extends LobbySectionBlock {
  final SuggestedGamesSectionBlock data;
  const SuggestedGamesSection(this.data);

  @override
  Map<String, dynamic> toJson() => data.toJson();
}

class GameGridSection extends LobbySectionBlock {
  final GameGridSectionBlock data;
  const GameGridSection(this.data);

  @override
  Map<String, dynamic> toJson() => data.toJson();
}

class BannerSection extends LobbySectionBlock {
  final BannerSectionBlock data;
  const BannerSection(this.data);

  @override
  Map<String, dynamic> toJson() => data.toJson();
}

/// JSON converter for List<LobbySectionBlock>
class LobbySectionBlockListConverter
    implements JsonConverter<List<LobbySectionBlock>, List<dynamic>> {
  const LobbySectionBlockListConverter();

  @override
  List<LobbySectionBlock> fromJson(List<dynamic> json) {
    return json
        .map((e) => LobbySectionBlock.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  @override
  List<dynamic> toJson(List<LobbySectionBlock> object) {
    return object.map((e) => e.toJson()).toList();
  }
}

@freezed
class LobbyLayout with _$LobbyLayout {
  const factory LobbyLayout({
    required String id,
    required String slug,
    required String name,
    required Platform platform,
    @Default(false) bool isDefault,
    @LobbySectionBlockListConverter() required List<LobbySectionBlock> sections,
    String? createdAt,
    String? updatedAt,
  }) = _LobbyLayout;

  factory LobbyLayout.fromJson(Map<String, dynamic> json) =>
      _$LobbyLayoutFromJson(json);
}
